proposals/20250923-plugin-system-event-schema-versioning.md

```diff
@@ -0,0 +1,169 @@
+# Proposal for the introduction of driver event schema version check in the Plugin System
+
+## Summary
+
+This proposal introduces a comprehensive event schema version validation system for the Falco plugin framework. The system will provide runtime validation of the event schema version related to the events generated by the drivers, that can be consumed by plugins providing parsing and/or field extraction capabilities. This will ensure backward compatibility and clear error reporting for plugins that depend on specific schema versions.
+
+## Motivation
+
+The current Falco plugin system only checks the Plugin API version, but there is no check on the driver event schema version. This means that there is no guarantee that a given plugin with parsing or field extraction capabilities works across Falco versions with different schema versions.
+
+1. **Runtime Validation**: There's no mechanism for plugins consuming driver events to declare the required schema version
+1. **Version Compatibility**: No way to enforce compatibility between plugins and the event schema version
+
+## Goals
+
+- Implement a comprehensive driver event schema validation system for plugins
+- Provide backward compatibility guarantees for schema evolution
+- Enable runtime schema validation with clear error reporting
+- Not breaking the existing plugin APIs (this should be additive)
+
+## Non-Goals
+
+- Guaranteeing interoperability between plugins using the plugin event type
+  (this is out of scope for now)
+
+## Proposal
+
+### 1. Required schema version declaration
+
+The plugins should declare the required schema version in the following cases.
+
+- field extraction capability, unless the plugin will only receive events of the plugin type
+- event parsing capability, unless the plugin will only receive events of the plugin type
+
+If the plugin does not declare a required schema version, it is assumed that it
+is compatible with 3.0.0, the initial major version at the time the plugin
+event schema validation is introduced.
```

Comment:
Since this validation check **is introduced at the same time as bumping the SCHEMA_VERSION to 4.0.0** (it is worth mentioning this in the proposal due to its relevance), the version check does not pass in any case (since `3` is always not equal to `>=4`).

Thus, we can simplify this by saying that the validation will always fail when a plugin does not declare a required schema version or declare a version minor of `4.0.0`.

Reply:
I did like this to stay consistent with the current state. The bump of the major schema version did not happen yet. Anyway, I clarified that 3.0.0 is the major version of the release that pre-dates the introduction of the check.